src/autora/experimentalist/random_.py

```diff
@@ -0,0 +1,341 @@
+
+
+        We can specify arrays of allowed values:
+
+        >>> random_pool(
+        ...     VariableCollection(independent_variables=[
+        ...         Variable(name="x", allowed_values=np.linspace(-10, 10, 101)),
+        ...         Variable(name="y", allowed_values=[3, 4]),
+        ...         Variable(name="z", allowed_values=np.linspace(20, 30, 11)),
+        ... ]), random_state=1)["conditions"]
+             x  y     z
+        0 -0.6  3  29.0
+        1  0.2  4  24.0
+        2  5.2  4  23.0
+        3  9.0  3  29.0
+        4 -9.4  3  22.0
+
+
+    """
+    rng = np.random.default_rng(random_state)
+
+    raw_conditions = {}
+    for iv in variables.independent_variables:
+        if iv.allowed_values is not None:
+            raw_conditions[iv.name] = rng.choice(
+                iv.allowed_values, size=num_samples, replace=replace
+            )
+        elif (iv.value_range is not None) and (iv.type == ValueType.REAL):
+            raw_conditions[iv.name] = rng.uniform(*iv.value_range, size=num_samples)
+
+        else:
+            raise ValueError(
+                "allowed_values or [value_range and type==REAL] needs to be set for "
+                "%s" % (iv)
+            )
+
+    conditions = pd.DataFrame(raw_conditions)
+    return Result(conditions=conditions)
+
+
+@singledispatch
+def random_sample(s, **kwargs):
+    """Function to create a sequence of conditions randomly sampled from conditions."""
+    raise NotImplementedError(
+        "random_sample doesn't have an implementation for %s (type=%s)" % (s, type(s))
+    )
+
+
+@random_sample.register(State)
+def random_sample_on_state(s: State, **kwargs) -> State:
+    return wrap_to_use_state(random_sample_on_conditions)(s, **kwargs)
+
+
+@random_sample.register(list)
+@random_sample.register(tuple)
+def random_sample_on_list(
+    conditions: Union[list, tuple],
+    num_samples: int = 1,
+    random_state: Optional[int] = None,
+    replace: bool = False,
+) -> list:
+    """
+    Examples:
+        >>> random_sample([1, 1, 2, 2, 3, 3], num_samples=2, random_state=1, replace=True)
+        [1, 3]
+
+        >>> random_sample((1, 1, 2, 2, 3, 3), num_samples=3, random_state=1, replace=True)
+        [1, 3, 3]
+
+
+    """
+
+    if random_state is not None:
+        random.seed(random_state)
+
+    assert replace is True, "random.choices only supports choice with replacement."
+    return random.choices(conditions, k=num_samples)
+
+
+@random_sample.register(pd.DataFrame)
+@random_sample.register(np.ndarray)
+@random_sample.register(np.recarray)
+def random_sample_on_conditions(
+    conditions: Union[pd.DataFrame, np.ndarray, np.recarray],
+    num_samples: int = 1,
+    random_state: Optional[int] = None,
+    replace: bool = False,
+) -> Result:
+    """
+    Take a random sample from some conditions.
+
+    Args:
+        conditions: the conditions to sample from
+        num_samples:
```

Comment:
same question as before; is there a reason for leaving these entries blank? seems like this one might just be an oversight, but less sure about `Returns` above

Reply:
same again – failure to notice. Updated now.